src/wkwebview/class/wry_download_delegate.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2020-2024 Tauri Programme within The Commons Conservancy
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
+
+use std::{path::PathBuf, ptr::null_mut, rc::Rc};
+
+use objc2::{
+  declare_class, msg_send_id, mutability::MainThreadOnly, rc::Retained, runtime::NSObject,
+  ClassType, DeclaredClass,
+};
+use objc2_foundation::{
+  MainThreadMarker, NSData, NSError, NSObjectProtocol, NSString, NSURLResponse, NSURL,
+};
+use objc2_web_kit::{WKDownload, WKDownloadDelegate};
+
+use crate::wkwebview::download::{download_did_fail, download_did_finish, download_policy};
+
+pub struct WryDownloadDelegateIvars {
+  pub started: *mut Box<dyn FnMut(String, &mut PathBuf) -> bool>,
+  pub completed: *mut Rc<dyn Fn(String, Option<PathBuf>, bool)>,
+}
+
+declare_class!(
+  pub struct WryDownloadDelegate;
+
+  unsafe impl ClassType for WryDownloadDelegate {
+    type Super = NSObject;
+    type Mutability = MainThreadOnly;
+    const NAME: &'static str = "WryDownloadDelegate";
+  }
+
+  impl DeclaredClass for WryDownloadDelegate {
+    type Ivars = WryDownloadDelegateIvars;
+  }
+
+  unsafe impl NSObjectProtocol for WryDownloadDelegate {}
+
+  unsafe impl WKDownloadDelegate for WryDownloadDelegate {
+    #[method(download:decideDestinationUsingResponse:suggestedFilename:completionHandler:)]
+    fn download_policy(
+      &self,
+      download: &WKDownload,
+      response: &NSURLResponse,
+      suggested_path: &NSString,
+      handler: &block2::Block<dyn Fn(*const NSURL)>,
+    ) {
+      download_policy(self, download, response, suggested_path, handler);
+    }
+
+    #[method(downloadDidFinish:)]
+    fn download_did_finish(&self, download: &WKDownload) {
+      download_did_finish(self, download);
+    }
+
+    #[method(download:didFailWithError:resumeData:)]
+    fn download_did_fail(
+      &self,
+      download: &WKDownload,
+      error: &NSError,
+      resume_data: &NSData,
+    ) {
+      download_did_fail(self, download, error, resume_data);
+    }
+  }
+);
+
+impl WryDownloadDelegate {
+  pub fn new(
+    download_started_handler: Option<Box<dyn FnMut(String, &mut PathBuf) -> bool>>,
+    download_completed_handler: Option<Rc<dyn Fn(String, Option<PathBuf>, bool)>>,
```

Comment:
If you change these to take `dyn Fn(...) + 'static` instead, then you don't need to `into_raw` and `from_raw` it, and it also becomes sound. (At least I think that's how the lifetimes work here? I can never remember...)

Also, `Box<dyn FnMut(...)>` really isn't possible in Objective-C classes, you'll have to wrap it in a `RefCell`.
```suggestion
    download_started_handler: Option<RefCell<Box<dyn FnMut(String, &mut PathBuf) -> bool + 'static>>>,
    download_completed_handler: Option<Rc<dyn Fn(String, Option<PathBuf>, bool) + 'static>>,
```

Reply:
> Also, Box<dyn FnMut(...)> really isn't possible in Objective-C classes, you'll have to wrap it in a RefCell.

Could you explain more about this? I am curious how this work

Reply:
Objective-C classes require interior mutability (`&self`), so while you _can_ store `Box<dyn FnMut(...)>` in them, it'd be useless, as it requires a mutable reference (`&mut self`) to call the closure.

It'd be kinda like using `Rc<Box<dyn FnMut(...)>>` - [try it out](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=45a96d1cbcf586342b8e7d0701627c11), it won't work unless you introduce `RefCell` or similar forms of interior mutability.

Just to clear up any confusion, `Box<dyn Fn(...)>` is fine, it's the `FnMut` that's problematic.